clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/BeanExpirationScheduler.java

```diff
@@ -38,46 +36,49 @@
  * Schedules a bean for expiration.
  *
  * @author Paul Ferraro
+ * @author Flavia Rainone
  *
  * @param <I> the bean identifier type
  * @param <T> the bean type
  */
 public class BeanExpirationScheduler<I, T> implements Scheduler<I> {
-    final Map<I, Future<?>> expirationFutures = new ConcurrentHashMap<>();
-    final Batcher<TransactionBatch> batcher;
-    final BeanRemover<I, T> remover;
-    final ExpirationConfiguration<T> expiration;
+    private final Batcher<TransactionBatch> batcher;
+    private final BeanRemover<I, T> remover;
+    private final ExpirationConfiguration<T> expiration;
+    private final ExpirationTracker<I> expirationTracker;
+    private Future<?> expireTask;
 
-    public BeanExpirationScheduler(Batcher<TransactionBatch> batcher, BeanRemover<I, T> remover, ExpirationConfiguration<T> expiration) {
+    BeanExpirationScheduler(Batcher<TransactionBatch> batcher, BeanRemover<I, T> remover, ExpirationConfiguration<T> expiration) {
         this.batcher = batcher;
         this.remover = remover;
         this.expiration = expiration;
+        this.expirationTracker = expiration.getTimeout().isNegative()? null : new ExpirationTracker<>(expiration.getTimeout());
     }
 
     @Override
     public void schedule(I id) {
         Duration timeout = this.expiration.getTimeout();
-        if (!timeout.isNegative()) {
-            InfinispanEjbLogger.ROOT_LOGGER.tracef("Scheduling stateful session bean %s to expire in %s", id, timeout);
-            Runnable task = new ExpirationTask(id);
-            // Make sure the expiration future map insertion happens before map removal (during task execution).
-            synchronized (task) {
-                this.expirationFutures.put(id, this.expiration.getExecutor().schedule(task, timeout.toMillis(), TimeUnit.MILLISECONDS));
+        if (expirationTracker != null) {
+            expirationTracker.trackExpiration(id);
+            if (expireTask == null) {
+                InfinispanEjbLogger.ROOT_LOGGER.tracef("Scheduling stateful session bean %s to expire in %s", id, timeout);
+                Runnable task = new ExpirationTask();
+                // Make sure the expiration future map insertion happens before map removal (during task execution).
+                synchronized (this) {
+                    expireTask = this.expiration.getExecutor().schedule(task, timeout.toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
I think you want an additional null check here - since multiple concurrent threads can reach this synchronized block resulting in multiple expire task creation.
e.g.
```java
synchronized (this) {
   if (expireTask == null) {
       expireTask = ...;
   }
}
```

Reply:
+1